### ISSUE TITLE
Use of EsiApp failing

### ISSUE THREAD

**Issue by EricE** [Kyria/EsiPy#32]
I'm trying to use EsiApp so that I get caching of the Swagger App object.  I'm following the example under Advanced Uses / EsiApp - Esi Meta Spec in your tutorial and I get the following error:

File "d:\SVNRepo\projects-sw\EVE_ESI\trunk\env\Lib\site-packages\pyswagger\utils.py", line 72, in __getitem__
  return super(ScopeDict, self).__getitem__(k)

builtins.KeyError: 'get_universe_system_kills'

My code:

    from esipy import EsiApp
    
    esi_app = EsiApp() # I'll add a diskcache when it is working
    security = EsiSecurity(
	    redirect_uri='xxx',
	    client_id='xxx',
	    secret_key='xxx',
    )
    
    client = EsiClient(
	    retry_requests=True,
	    header={'User-Agent': 'ESI testing'},
	    security=security
    )
 
((authorize here))

This fails with the above error:
    op = esi_app.op\['get_universe_system_kills'\]()
    
This also fails with the above error:
    op = esi_app.app.op\['get_universe_system_kills'\]()
    
To get a result I have to do:

    esi_current = esi_app.get_latest_swagger
    op = esi_current.op['get_universe_system_kills']()

But I lose the benefit of caching the Swagger App object.

I don't understand why your __get_or_create_app version of App is different than the one created by:

    from esipy import App
    App.create(url="https://esi.tech.ccp.is/latest/swagger.json") in my code which works...

**Comment by Kyria** [comment]
tl;dr;
---

This is the intended behavior, you actually should `esi_app.get_<version>_swagger.op['get_universe_system_kills']()` and not `esi_app.op['get_universe_system_kills']()`

EsiApp is mostly a shortcut (with cache management, unless disabled) for every version of ESI, so you don't have to create and manage caching for theses endpoints. 

I hope i've answered your issue with the detailed version below. 
---

Detailed version
---

This is because it's the intended behavior: EsiApp is here to deal with the meta endpoints, allowing us to fetch any version easily while caching them at the same time: 

```python
esi_latest = esi_app.get_latest_swagger
esi_v1 = esi_app.get_v1_swagger
esi_v2 = esi_app.get_v2_swagger
esi_v3 = esi_app.get_v3_swagger
....

esi_v1_again = esi_app.get_v1_swagger # won't fetch from web but use cache
```

The operation allowed with EsiApp are the one here: https://esi.tech.ccp.is/ui/?version=meta so you can do `esi_app.op['something']()` but only on meta operations, like `verify`. 



> I don't understand why your __get_or_create_app version of App is different than the one created by:
>`from esipy import App`
>`App.create(url="https://esi.tech.ccp.is/latest/swagger.json") in my code which works...`

`App.create(url="https://esi.tech.ccp.is/latest/swagger.json")` this is equivalent to `esi_app.get_latest_swagger` only. <br>
`__get_or_create_app` adds just a layer above to manage caching, so the end result is the same. 

>But I lose the benefit of caching the Swagger App object.

 
Actually, you don't lose the caching by doing 
```
esi_current = esi_app.get_latest_swagger
op = esi_current.op['get_universe_system_kills']()
```
Because:
- esi_app caches the meta endpoint and individual versioned endpoints
- each individual esi_current, once they are created, do not update anything unless you recreate them. 

Also things like `esi_app.get_latest_swagger` will first check the cache before fetching from web, so you can do things like:
```
esi_current = esi_app.get_latest_swagger
op = esi_current.op['get_universe_system_kills']()

# some long work 
# let's say 10k line of code

esi_current = esi_app.get_latest_swagger
op = esi_current.op['another_endpoint']()
```
this will use the cache within esi_app, unless you manually clear everything, or the data are outdated.

**Comment by EricE** [comment]
I really appreciate it that you take the time for such detailed responses.  I turned on file caching and now I see that it is caching the various versions.  Before, I was trying to single step through in my IDE (Wing) looking for the cache check and it was erroring out at:

in EsiApp
attr = super(EsiApp, self).__getattribute__(name)

so I set more breakpoints and I see now where it is checking the cache.

I went back through the dev blogs and I see why you say not to use latest:

"The short version: always use the versioned routes in your software, never use labels. Labels can change, versions do not."

However since the routes are all at different versions, figuring out which version of get_vX_swagger to use seems complicated.  I tried to just grab the latest, i.e. get_v4_swagger, but it appears that it only understands routes at that version.  It seems like the logic to match the route version to the swagger app version would be tricky, do you have a recommended best practice? I looked at LazyBlacksmith but it looks like it is using latest?

Also, what is the best way to get the latest version of a route?  I tried 

    esi_latest.op['get_characters_character_id_orders'].dump() 

but v2 only shows up in the description.  I also saw a comment in a dev blog that implied there is no way to get the versions directly.  I'll add checks for 'Warning' in op[1].header (better way?) as a runtime check but also think that it would be good as a daily task to scan for updates (which I guess could also check for the warning).

**Comment by Kyria** [comment]
use this one https://esi.tech.ccp.is/ui/?version=_latest (`get__latest_swagger`): it's the same as latest, but shows version instead of "latest", so you know what you use, and you can fix it after :) 

About LB, that's fun, i was sure I already changed it... but nop. About best practice, I can't recommend anything, but what i'd do (and what I thought I committed for LB) is to use each version, either by `esiapp.get_vX_swagger.op['somethingsomething']` or by saving them all and using the "shortcut" (as I do in LB).

For versions, yes, each version only shows endpoints that have that version, so your best bet is to find the latest on `_latest` then fix it in the code to use the given version. In the latest, you can't find the version yes, unless it's in the description, also the warning won't ever show if you use latest, as it only show for versions within "legacy" (so not anymore in "latest"). 

You'd better fix the version, check for warning, and update your code if there's a warning :)

**Comment by EricE** [comment]
Thanks again for all the help, it got me over the rough spot and I'm off and running.
